corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,924 @@
+    check.real_positive_scalar(actual_gain_mean_frame, 'actual_gain_mean_frame', TypeError)
+    if actual_gain_mean_frame < 1:
+        raise CalKgainException('Actual gain must be >= 1.')
+    check.positive_scalar_integer(min_val, 'min_val', TypeError)
+    check.positive_scalar_integer(max_val, 'max_val', TypeError)
+    if min_val >= max_val:
+        raise CalKgainException('max_val must be greater than min_val')
+    check.positive_scalar_integer(binwidth, 'binwidth', TypeError)
+    if binwidth < 10:
+        raise CalKgainException('binwidth must be >= 10.')
+    if binwidth > 800:
+        raise CalKgainException('binwidth must be < 800.')
+    if not isinstance(logspace_start, (float, int)):
+        raise TypeError('logplot1 is not a number')
+    if not isinstance(logspace_stop, (float, int)):
+        raise TypeError('logplot2 is not a number')
+    if not isinstance(logspace_num, (float, int)):
+        raise TypeError('logplot3 is not a number')
+    
+    # get relevant constants
+    offset_rowroi1 = kgain_params['offset_rowroi1']
+    offset_rowroi2 = kgain_params['offset_rowroi2']
+    offset_colroi1 = kgain_params['offset_colroi1']
+    offset_colroi2 = kgain_params['offset_colroi2']
+    rowroi1 = kgain_params['rowroi1']
+    rowroi2 = kgain_params['rowroi2']
+    colroi1 = kgain_params['colroi1']
+    colroi2 = kgain_params['colroi2']
+    rn_bins1 = kgain_params['rn_bins1']
+    rn_bins2 = kgain_params['rn_bins2']
+    max_DN_val = kgain_params['max_DN_val']
+    signal_bins_N = kgain_params['signal_bins_N']
+
+    if make_plot is True:
+        # Avoid issues with importing matplotlib on headless servers without GUI
+        # support without proper configuration
+        import matplotlib.pyplot as plt
+        # Output directory
+        if os.path.exists(plot_outdir) is False:
+            os.mkdir(plot_outdir)
+            if verbose:
+                print('Output directory for figures created in ', os.getcwd())
+    
+    # Prescan region
+    offset_rowroi = slice(offset_rowroi1,offset_rowroi2)
+    offset_colroi = slice(offset_colroi1,offset_colroi2)
+    
+    # NOTE: binwidth must be <= rowroi and colroi
+    rowroi = slice(rowroi1,rowroi2)
+    colroi = slice(colroi1,colroi2)
+    
+    # ROI for frameProc
+    colroi_fp = slice(offset_colroi1,offset_colroi2)
+    
+    averages = []
+    deviations = []
+    read_noise = []
+    deviations_shot = []
+    #mean_signals = []
+    
+    rn_bins = np.array(range(rn_bins1, rn_bins2))
+    bin_locs = rn_bins[0:-1]
+
+    max_DN = max_DN_val; # maximum DN value to be included in PTC
+    nrow = len(cal_list[0][0])
+    ncol = len(cal_list[0][0][0])
+    
+    # prepare "good mean frame"
+    good_mean_frame = np.zeros((nrow, ncol))
+    nFrames2 = len(mean_frame_list)
+    for mean_frame_count in range(nFrames2):
+        frame = mean_frame_list[mean_frame_count]
+        frame = frameProc(frame, colroi_fp, actual_gain)
```

Comment:
The `frameProc ` function is superseded by the `prescan_biassub` step function in l1_to_l2a.py and the `em_gain_division` step function in `l2a_to_l2b.py`. I think it would be fair to make it a requirement that the input dataset has already been processed by these steps.

Reply:
Last thing I heard is that it could be mentioned in the doc strings, as it is now. I'm fine changing it to not allow it and expect it to be processed earlier on.

Reply:
Yes, if we had to I suppose we could just have this function accept L1 data. It ends up bypassing much of the header tracking of frames, which is not ideal and means that if we ever change anything about either the EMGAIN or the pre-scan bias routines we'd have to change them here too, so it's really not ideal from a code maintenance point of view. I'm not sure I understand the advantage of doing it this way.

Reply:
I'm not advocating for it. I was just paraphrasing the decision on the meeting. I'll implement the idea that bias and EM gain division are already applied. Maybe @kjl0025 has some objection/suggestion not to do it?

Reply:
I agree with Max here. We should defer these steps to the other step functions we wrote.

Reply:
I agree that the frames should be assumed to be already processed as much as possible before going through this function.  In the case of calibrate_kgain and calibrate_nonlin, the frames are bias-subtracted, and in addition, if EM gain is >1 for the input data for calibrate_kgain, EM gain division is also needed.  So the bias subtraction that happens here (and in the nonlin code) needs to be removed, along with the EM gain division in this function.  Consequently, we don't need to worry about changing the bias subtraction implementation here to use `slice_section`, and thus we also don't need offset_rowroi1, offset_rowroi2, offset_colroi1, or offset_colroi2. And the doc string should be changed accordingly to specify that these pipeline steps should have already been taken for the input data.

Reply:
@kjl0025: "don't need offset_rowroi1, offset_rowroi2, offset_colroi1, or offset_colroi2." These are used in calibrate_kgain.py for other functions than the bais subtraction. Similarly for calibrate_nonlin.py. I think I have to keep them and remove frameProc, change doc strings and inout bias subtracted (EM divided, in Kgain) in the unit tests

Reply:
I have committed the changes above, except removing "ffset_rowroi1, offset_rowroi2, offset_colroi1, or offset_colroi2" for the reasons I mention. If there's something else to do about them, let me know. Otherwise, we can resolve this converstaion.

Reply:
Oh, true.  It looks like the offset_* parameters are only used for specifying the prescan region (other than bias subtraction via frameProc).  We can specify the prescan portion of a frame instead using `slice_section`.  But as discussed before, if that is being put off to a future pull request, then you'll need to keep the offset_* parameters for now.  I don't think that change would be hard to do.  Anywhere you see a frame region is specified as `frame[offset_rowroi, offset_colroi]`, you could instead use `prescan = slice_section(frame, 'SCI', 'prescan_reliable', detector_regions)` where `detector_regions` is as found in `prescan_biassub()`.

Reply:
Done.

Thanks so much everyone and kudos for @kjl0025 and @maxwellmb for all the extensive details! 

Please, let me know if you have any further suggestions!